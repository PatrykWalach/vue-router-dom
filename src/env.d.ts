/// <reference types="vue/macros-global" />

declare module '*.vue' {
  import type { Component } from 'vue'
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/ban-types
  const component: Component<{}, {}, any>
  export default component
}
