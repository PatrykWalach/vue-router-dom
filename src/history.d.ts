declare module 'history' {
  interface HashLocation {
    pathname: string
    search: string
    hash: string
  }

  interface Location extends HashLocation {
    state: any
    key: string
  }

  type Action = 'PUSH' | 'REPLACE' | 'POP'

  type Listener = (location: Location, action: Action) => any
  type HashListener = (location: HashLocation, action: Action) => any

  type Listen<Listener> = (listener: Listener) => HistoryUnlisten
  type HistoryUnlisten = () => void

  interface Path {
    pathname?: string
    search?: string
    hash?: string
    state?: any
  }

  interface BaseHistory {
    push(path: Path): void
    push(path: string, state?: any): void
    replace(path: Path): void
    replace(path: string, state?: any): void
    go(n: number): void
    goBack(): void
    goForward(): void
    length: number
    action: Action
  }

  interface HashHistory extends BaseHistory {
    location: HashLocation
    listen: Listen<HashListener>
  }

  interface BrowserHistory extends BaseHistory {
    location: Location
    listen: Listen<Listener>
  }

  interface MemoryHistory extends BrowserHistory {
    canGo(n: number): void
    index: number
    entries: string[]
  }

  interface BrowserHistoryOptions {
    basename?: string
    forceRefresh?: boolean
    keyLength?: number
    getUserConfirmation?: Function
  }

  interface HashHistoryOptions {
    basename?: string
    hashType?: string
    getUserConfirmation?: Function
  }

  interface MemoryHistoryOptions {
    initialEntries?: string[]
    initialIndex?: number
    keyLength?: number
    getUserConfirmation?: Function
  }

  type CreateBrowserHistory = (
    options?: BrowserHistoryOptions,
  ) => BrowserHistory
  type CreateHashHistory = (options?: HashHistoryOptions) => HashHistory
  type CreateMemoryHistory = (options?: MemoryHistoryOptions) => MemoryHistory

  export const createBrowserHistory: CreateBrowserHistory
  export const createHashHistory: CreateHashHistory
  export const createMemoryHistory: CreateMemoryHistory
}
