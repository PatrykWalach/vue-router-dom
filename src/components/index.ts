export { default as BrowserRouter } from './BrowserRouter.vue'
export { default as Link } from './Link.vue'
export { default as MemoryRouter } from './MemoryRouter.vue'
export { default as NavLink } from './NavLink.vue'
export { default as Outlet } from './Outlet.vue'
export { default as Route } from './Route.vue'
export { default as Router } from './Router.vue'
export { default as Routes } from './Routes.vue'
