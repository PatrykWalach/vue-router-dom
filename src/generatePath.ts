import { PathFunction, compile } from 'path-to-regexp'
import { RouterParams } from './types'

const cache = new Map<string, PathFunction<object>>()

const cacheLimit = 10000
let cacheCount = 0

export const generatePath = (pattern: string, params: RouterParams) => {
  const cached = cache.get(pattern)
  if (cached) return cached

  const toPath = compile(pattern, { encode: encodeURIComponent })

  if (cacheCount < cacheLimit) {
    cache.set(pattern, toPath)
    cacheCount++
  }

  return toPath(params)
}
