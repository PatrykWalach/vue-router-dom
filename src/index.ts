export * from './remix/keys'
export * from './remix/types'
export * from './remix/hooks'
export { default as DataBrowserRouter } from './components/DataBrowserRouter.vue'
export { default as DataMemoryRouter } from './components/DataMemoryRouter.vue'
export { default as BrowserRouter } from './components/BrowserRouter.vue'
export { default as MemoryRouter } from './components/MemoryRouter.vue'
export { default as Link } from './components/Link.vue'
export { default as Outlet } from './components/Outlet.vue'
export { default as Routes } from './components/Routes.vue'

export { useRoutes } from './remix/useRoutes'

export type { VueRouteObject as RouteObject } from './remix/types'
