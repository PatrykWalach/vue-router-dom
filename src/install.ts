import { App } from 'vue'
//install temporarly disabled because of type error

// import { BrowserRouter } from './components/BrowserRouter'
// import { RouterLink } from './components/RouterLink'
// import { RouterRoute } from './components/RouterRoute'
// import { RouterSwitch } from './components/RouterSwitch'

export const install = (app: App) => app
// .component('BrowserRouter', BrowserRouter)
// .component('RouterLink', RouterLink)
// .component('RouterRoute', RouterRoute)
// .component('RouterSwitch', RouterSwitch)
