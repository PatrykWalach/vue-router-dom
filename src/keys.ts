import { InjectionKey, Ref } from 'vue'
import { RouterHistory, RouterMemoryHistory } from 'history'
import { RouterParams } from './types'

export const ROUTE_PARAMS: InjectionKey<Readonly<Ref<RouterParams>>> = Symbol(
  'ROUTER_HISTORY',
)

export const ROUTER_HISTORY: InjectionKey<
  RouterHistory | RouterMemoryHistory
> = Symbol('ROUTER_HISTORY')
