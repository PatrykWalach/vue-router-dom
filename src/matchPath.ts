import { Key, pathToRegexp } from 'path-to-regexp'
import {
  MatchPathOptions,
  MatchPathOptionsPath,
  RouterMatch,
  RouterParams,
} from './types'

interface CompilePathOptions {
  end: boolean
  strict: boolean
  sensitive: boolean
}
interface CompilePathResult {
  regexp: RegExp
  keys: Key[]
}

const cache = new Map<string, Map<string, CompilePathResult>>()

const cacheLimit = 10000
let cacheCount = 0

const compilePath = (
  path: string,
  options: CompilePathOptions,
): CompilePathResult => {
  const cacheKey = `${options.end}${options.strict}${options.sensitive}`

  const pathCache = cache.get(cacheKey) || new Map<string, CompilePathResult>()

  cache.set(cacheKey, pathCache)

  const cached = pathCache.get(path)
  if (cached) return cached

  const keys: Key[] = []
  const regexp = pathToRegexp(path, keys, options)
  const result = { keys, regexp }

  if (cacheCount < cacheLimit) {
    pathCache.set(path, result)
    cacheCount++
  }

  return result
}

export const matchPath = <P extends RouterParams = RouterParams>(
  pathname: string,
  rawOptions: MatchPathOptions | MatchPathOptionsPath = {},
) => {
  const options: MatchPathOptions =
    typeof rawOptions === 'string' || Array.isArray(rawOptions)
      ? { path: rawOptions }
      : rawOptions

  const { path, exact = false, strict = false, sensitive = false } = options

  const paths = path instanceof Array ? path : [path]

  return paths.reduce<RouterMatch<P> | null>((matched, path) => {
    if (!path && path !== '') {
      return null
    }
    if (matched) {
      return matched
    }

    const { regexp, keys } = compilePath(path, {
      end: exact,
      sensitive,
      strict,
    })
    const match = regexp.exec(pathname)

    if (!match) {
      return null
    }

    const [url, ...values] = match
    const isExact = pathname === url

    if (exact && !isExact) {
      return null
    }

    return {
      isExact,
      params: keys.reduce<RouterParams>((memo, key, index) => {
        memo[key.name] = values[index]
        return memo
      }, {}),
      path,
      url: path === '/' && url === '' ? '/' : url,
    } as RouterMatch<P>
  }, null)
}
