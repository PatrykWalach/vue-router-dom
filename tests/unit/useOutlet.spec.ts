import { Routes, Route, MemoryRouter, Outlet } from '../../src'

import { defineComponent, h } from 'vue'
import { useOutlet } from '../../src'
import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
describe('useOutlet', () => {
  describe('when there is no child route', () => {
    it('returns null', () => {
      const wrapper = mount({
        render: () =>
          h(MemoryRouter, { initialEntries: ['/home'] }, () =>
            h(Routes, {
              routes: [{ path: '/home', element: Outlet }],
            }),
          ),
      })

      expect(wrapper.html()).toMatchInlineSnapshot(`""`)
    })
  })

  describe('when there is a child route', () => {
    it('returns an element', () => {
      const Profile = () => h('p', 'Profile')

      const wrapper = mount({
        render: () =>
          h(MemoryRouter, { initialEntries: ['/users/profile'] }, () =>
            h(Routes, {
              routes: [
                {
                  path: 'users',
                  element: Outlet,
                  children: [{ path: 'profile', element: Profile }],
                },
              ],
            }),
          ),
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
