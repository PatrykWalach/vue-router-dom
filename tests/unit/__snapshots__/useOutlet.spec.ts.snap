// Vitest Snapshot v1

exports[`useOutlet > when there is a child route > returns an element 1`] = `"<p>Profile</p>"`;

exports[`useOutlet when there is a child route returns an element 1`] = `"<p>Profile</p>"`;
